GameProject/Assets/TurnBasedAssets/Scripts/PathFinding/Pathfinder.cs

```diff
@@ -0,0 +1,146 @@
+                {
+                    break;
+                }
+
+                adjacentSquares.Clear();
+                adjacentSquares = !is3d ? GetAdjacentSquares2D(currentLocation, targetPosition, isObjectMoving) : GetAdjacentSquares3D(currentLocation, targetPosition, isObjectMoving);
+
+                foreach (var adjacentSquare in adjacentSquares)
+                {
+                    if (closedList.Any(x => x.PositionInWorld == adjacentSquare.PositionInWorld)) continue;
+
+                    if (openList.All(x => x.PositionInWorld != adjacentSquare.PositionInWorld)) openList.Add(adjacentSquare);
+
+                    else if (adjacentSquare.F < openList.First(x => x.PositionInWorld == adjacentSquare.PositionInWorld).F) openList.First(x => x.PositionInWorld == adjacentSquare.PositionInWorld).Parent = adjacentSquare;
+                }
+
+                yield return null;
+            }
+
+            _pathToFollow.Clear();
+
+            var current = closedList.Last();
+
+            _pathToFollow.Add(current.PositionInWorld);
+            
+            do
+            {
+                _pathToFollow.Add(current.Parent.PositionInWorld);
+                current = current.Parent;
+            } while (!_pathToFollow.Contains(startPosition));
+
+            _pathToFollow.Reverse();
+
+            onCompletion(_pathToFollow);
+        }
+
+        private List<Location> GetAdjacentSquares2D(Location point, Vector3 target, Controller isObjectMoving)
+        {
+            List<Location> returnList = new List<Location>();
+
+            for (float xIndex = point.PositionInWorld.x - 1; xIndex <= point.PositionInWorld.x + 1; xIndex++)
+            {
+                for (float zIndex = point.PositionInWorld.z - 1; zIndex <= point.PositionInWorld.z + 1; zIndex++)
+                {
+                    var adjacentVector =
+                        new Location(new Vector3(xIndex, point.PositionInWorld.y, zIndex), target, point);
+                    if (Vector3.Distance(point.PositionInWorld, adjacentVector.PositionInWorld) >
+                        _2dMaxDistance) continue;
+
+                    bool isIntersecting = _avoidance.Objects
+                        .Where(x => x!= isObjectMoving && Vector3.Distance(x.transform.position, adjacentVector.PositionInWorld) <=
+                                    Vector3.Distance(point.PositionInWorld, target)).Any(x => x.RenderBounds.bounds.Contains(adjacentVector.PositionInWorld));
+                    if (!isIntersecting) returnList.Add(adjacentVector);
+                }
+            }
+            return returnList;
+        }
+        
+        private List<Location> GetAdjacentSquares3D(Location point, Vector3 target, Controller isObjectMoving)
+        {
+            List<Location> returnList = new List<Location>();
+
+            for (float xIndex = point.PositionInWorld.x - 1; xIndex <= point.PositionInWorld.x + 1; xIndex++)
```

Comment:
While I understand why you've used a float as an iterator in this context it does read a little strange. For future note I advise using `int`. There's no point refactoring this now however as I can see it'd cause too many changes.